### DISCUSSION TITLE
PostCSS plugin outputs excessive warnings even when configured correctly

### DISCUSSION THREAD

**Discussion by Uninen** [tailwindlabs/tailwindcss#9487] [Help]
**What version of Tailwind CSS are you using?**

v3.1.8

**What build tool (or framework if it abstracts the build tool) are you using?**

Vite 3.1.4

```
    "@tailwindcss/nesting": "0.0.0-insiders.565cd3e",
    "autoprefixer": "10.4.12",
    "postcss": "8.4.17",
    "postcss-import": "15.0.0",
    "postcss-nesting": "10.2.0",
    "tailwindcss": "3.1.8",
    "vite": "3.1.4",
```

**What version of Node.js are you using?**

v16.15.1

**What browser are you using?**

N/A

**What operating system are you using?**

macOS

**Reproduction URL**

https://github.com/Uninen/ckeditor5-custom-vite

**Describe your issue**

I'm trying to build a custom CKEditor5 build w/ Vite *and* get Tailwind styles working in the same project. CKEditor config needs some wrangling because it assumes Webpack. Yet, it works, but only cleanly when Tailwind isn't configured at the same time.

When adding Tailwind config to an otherwise working CKEditor5 build project, the Tailwind plugin outputs a ton of following warnings (even though all the styles --including Tailwind ones-- are being built just fine):

```
vite:css] Nested CSS was detected, but CSS nesting has not been configured correctly.
Please enable a CSS nesting plugin *before* Tailwind in your configuration.
See how here: https://tailwindcss.com/docs/using-with-preprocessors#nesting
5  |  
6  |  .ck-content {
7  |    & .image {
   |    ^
8  |            display: table;
9  |            clear: both;
[vite:css] Nested CSS was detected, but CSS nesting has not been configured correctly.
Please enable a CSS nesting plugin *before* Tailwind in your configuration.
See how here: https://tailwindcss.com/docs/using-with-preprocessors#nesting
14 |    box-sizing: border-box;
15 |  
16 |    & img {
   |    ^
17 |            /* For resized images it is the `<figure>` element that determines the image width. */
18 |            width: 100%;
[vite:css] Nested CSS was detected, but CSS nesting has not been configured correctly.
Please enable a CSS nesting plugin *before* Tailwind in your configuration.
See how here: https://tailwindcss.com/docs/using-with-preprocessors#nesting
5  |  
6  |  .ck.ck-image-insert-form {
7  |    &:focus {
   |    ^
8  |            /* See: https://github.com/ckeditor/ckeditor5/issues/4773 */
9  |            outline: none;
```

I've tried all possible permutations for the ordering of the plugins. Either the styles don't work or I get this output from Tailwind (which makes Vite really slow and developing pretty impossible). Here's the config: https://github.com/Uninen/ckeditor5-custom-vite/blob/main/postcss.config.js

The Ckeditor config uses postcss-nesting, among others, full listing here: https://github.com/ckeditor/ckeditor5-dev/blob/master/packages/ckeditor5-dev-utils/lib/styles/getpostcssconfig.js

(The config itself is built from the comments in this ticket: https://github.com/ckeditor/ckeditor5/issues/9807)

**Comment by adamwathan** [comment]
The source of the issue here is that for whatever reason, you still have nested CSS in your CSS by the time it hits Tailwind. Any nesting has to be processed before Tailwind sees your CSS, as mentioned in the docs here: 

<img width="835" alt="image" src="https://user-images.githubusercontent.com/4323180/194156657-7aecaee1-83ac-4182-ad2c-0e526c3aa3bd.png">

It might be tricky with that CK editor stuff because they are using `postcss-nesting` directly which uses the newer PostCSS visitor API, and that API doesn't give you control over what order things run in, which is why we had to wrap it ourselves as a new plugin (`@tailwindcss/nesting`).

This isn't a bug in Tailwind though so I've converted it to a discussion and with any luck you or someone else can figure out how you need to configure things to avoid this conflict. You may want to look into wrapping the CK editor PostCSS config with your own PostCSS plugin that forces it to run first the same way we do with the nesting plugin.

**Comment by Uninen** [comment]
Ideally I'd like to disable that CSS (which comes from a third-party package, is pure PostCSS and has nothing to do with Tailwind) altogether somehow so Tailwind wouldn't even try to parse it. It just slows things down.

> You may want to look into wrapping the CK editor PostCSS config with your own PostCSS plugin that forces it to run first the same way we do with the nesting plugin.

Ok, that might be the last resort. Thanks for the tip 👍

**Comment by Manc** [comment]
I'm facing a similar issue. I followed the steps for [Create React App](https://tailwindcss.com/docs/guides/create-react-app) a few weeks ago and everything seemed fine. Either I didn't notice the error at first, but now I'm getting this:

```text
Warning
(45:2) Nested CSS was detected, but CSS nesting has not been configured correctly.
Please enable a CSS nesting plugin *before* Tailwind in your configuration.
See how here: https://tailwindcss.com/docs/using-with-preprocessors#nesting

Warning
(18:2) Nested CSS was detected, but CSS nesting has not been configured correctly.
Please enable a CSS nesting plugin *before* Tailwind in your configuration.
See how here: https://tailwindcss.com/docs/using-with-preprocessors#nesting

Warning
(2:2) Nested CSS was detected, but CSS nesting has not been configured correctly.
Please enable a CSS nesting plugin *before* Tailwind in your configuration.
See how here: https://tailwindcss.com/docs/using-with-preprocessors#nesting
```

Adding `'tailwindcss/nesting': {}` to my `postcss.config.js` didn't make a difference. In fact, I figured out, that this file seems to be ignored completely. No matter what I put in it (even invalid code), the result is always the same. My `tailwind.config.js` on the other hand, which is in the same directory, is being read correctly, including all customisations I made in there.

I have installed the latest package versions of `tailwindcss` and `postcss`.

**Reply by adamwathan** [reply]
CRA doesn’t let you touch the PostCSS config unfortunately so you’re sort of out of luck. I would highly recommend not using CRA, Vite is a much better option for the same type of use cases.
